src/util/timer.h

```diff
@@ -52,61 +57,60 @@ class SuspendableTimer : public Timer {
+                "means of u\"key text\"_s");
+        if (!CmdlineArgs::Instance().getDeveloper()) {
+            return; // leave timer in cancelled state
         }
-    }
-
-    ScopedTimer(const char* key, const char *arg = NULL,
-                Stat::ComputeFlags compute = kDefaultComputeFlags)
-            : m_pTimer(NULL),
-              m_cancel(false) {
-        if (CmdlineArgs::Instance().getDeveloper()) {
-            initialize(QString(key), arg ? QString(arg) : QString(), compute);
+        // key is explicitly `auto` to allow passing non-QString types such as `const char*`
+        // its conversion is delayed until after we've checked if we're in developer mode
+        auto assembledKey = QString(std::forward<T>(key));
+        if constexpr (sizeof...(args) > 0) {
+            // only try to call QString::arg when we've been given parameters
+            assembledKey = assembledKey.arg(convertToQStringConvertible(std::forward<Ts>(args))...);
         }
+        m_timer = std::make_optional<Timer>(assembledKey, compute);
+        m_timer->start();
     }
-
-    ScopedTimer(const char* key, const QString& arg,
-                Stat::ComputeFlags compute = kDefaultComputeFlags)
-            : m_pTimer(NULL),
-              m_cancel(false) {
-        if (CmdlineArgs::Instance().getDeveloper()) {
-            initialize(QString(key), arg, compute);
-        }
+    template<typename T, typename... Ts>
+    ScopedTimer(T&& key, Ts&&... args)
+            : ScopedTimer(kDefaultComputeFlags, std::forward<T>(key), std::forward<Ts>(args)...) {
     }
 
-    virtual ~ScopedTimer() {
-        if (m_pTimer) {
-            if (!m_cancel) {
-                m_pTimer->elapsed(true);
-            }
-            m_pTimer->~Timer();
+    ~ScopedTimer() {
+        if (m_timer.has_value()) {
+            m_timer->elapsed(true);
         }
     }
 
-    inline void initialize(const QString& key, const QString& arg,
-                Stat::ComputeFlags compute = kDefaultComputeFlags) {
-        QString strKey;
-        if (arg.isEmpty()) {
-            strKey = key;
-        } else {
-            strKey = key.arg(arg);
-        }
-        m_pTimer = new(m_timerMem) Timer(strKey, compute);
-        m_pTimer->start();
-    }
-
     void cancel() {
-        m_cancel = true;
+        m_timer.reset();
     }
   private:
-    Timer* m_pTimer;
-    char m_timerMem[sizeof(Timer)];
-    bool m_cancel;
+    // utility function
+    template<typename T>
+    constexpr static std::size_t char_type_size() {
+        if constexpr (std::is_array_v<T>) {
+            return sizeof(std::remove_all_extents_t<T>);
+        } else if constexpr (std::is_pointer_v<T>) {
+            // assume this is some point to char array then.
+            return sizeof(std::remove_pointer_t<T>);
+        } else {
+            // assume this is some QString or std::string type
+            return sizeof(typename T::value_type);
+        }
+    }
```

Comment:
I don't like how brittle this is tbh. I'd much rather just forbid the specific `const char*` so people don't accidentally do the inefficient `ScopedTimer t("key")` case and be done with it.

Reply:
It also doesn't work properly on the current CI. I could probably work around it, but I don't think its worth it.